src/NHibernate/Linq/Functions/DecimalGenerator.cs

```diff
@@ -0,0 +1,188 @@
+﻿using System.Collections.ObjectModel;
+using System.Linq.Expressions;
+using System.Reflection;
+
+using NHibernate.Hql.Ast;
+using NHibernate.Linq.Visitors;
+using NHibernate.Util;
+
+namespace NHibernate.Linq.Functions
+{
+	public class DecimalAddGenerator : BaseHqlGeneratorForMethod
+	{
+		public DecimalAddGenerator()
+		{
+			SupportedMethods = new[]
+			{
+				ReflectHelper.GetMethodDefinition(() => decimal.Add(default(decimal), default(decimal)))
+			};
+		}
+
+		public override HqlTreeNode BuildHql(MethodInfo method, Expression targetObject, ReadOnlyCollection<Expression> arguments, HqlTreeBuilder treeBuilder, IHqlExpressionVisitor visitor)
+		{
+			return treeBuilder.TransparentCast(treeBuilder.Add(visitor.Visit(arguments[0]).AsExpression(), visitor.Visit(arguments[1]).AsExpression()), typeof(decimal));
+		}
+	}
+
+	public class DecimalCompareGenerator : BaseHqlGeneratorForMethod
+	{
+		public DecimalCompareGenerator()
+		{
+			SupportedMethods = new[]
+			{
+				ReflectHelper.GetMethodDefinition(() => decimal.Compare(default(decimal), default(decimal)))
+			};
+		}
+
+		public override HqlTreeNode BuildHql(MethodInfo method, Expression targetObject, ReadOnlyCollection<Expression> arguments, HqlTreeBuilder treeBuilder, IHqlExpressionVisitor visitor)
+		{
+			return treeBuilder.MethodCall("sign", treeBuilder.Subtract(visitor.Visit(arguments[0]).AsExpression(), visitor.Visit(arguments[1]).AsExpression()));
+		}
+	}
+
+	public class DecimalDivideGenerator : BaseHqlGeneratorForMethod
+	{
+		public DecimalDivideGenerator()
+		{
+			SupportedMethods = new[]
+			{
+				ReflectHelper.GetMethodDefinition(() => decimal.Divide(default(decimal), default(decimal)))
+			};
+		}
+
+		public override HqlTreeNode BuildHql(MethodInfo method, Expression targetObject, ReadOnlyCollection<Expression> arguments, HqlTreeBuilder treeBuilder, IHqlExpressionVisitor visitor)
+		{
+			return treeBuilder.TransparentCast(treeBuilder.Divide(visitor.Visit(arguments[0]).AsExpression(), visitor.Visit(arguments[1]).AsExpression()), typeof(decimal));
```

Comment:
Putting `Cast` here fix Oracle failure. So long for avoiding having it in SQL for the other DBs which do not need it, provided they support having the cast.